### PR TITLE
move 'return outside method definition' to error class

### PR DESCRIPTION
**Pull request by ennru**
@felixmulder Long time, no commit.

### CODE REVIEW

**Comment by nicolasstucki** [comment]
@enru Will need to be rebased. Conflicts with the merge of your other PR

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 1]
```diff
@@ -1737,4 +1737,12 @@ object messages {
       hl"you have to provide either ${"class"}, ${"trait"}, ${"object"}, or ${"enum"} definitions after qualifiers"
   }
 
+  case class ReturnOutsideMethodDefinition()(implicit ctx: Context)
+    extends Message(ReturnOutsideMethodDefinitionID) {
+    val kind = "Syntax"
+    val msg = s"return outside method definition"
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: I would turn the phrasing around and say that `retrun` must be inside a method definition.

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 2]
```diff
@@ -1737,4 +1737,12 @@ object messages {
...
+    val explanation =
+      hl"${"return"} is a keyword and may only be used within methods"
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Here I would explain where the return vas found (i.e. in `ctx.owner` if it is not `NoContext`)

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala (hunk 1 above)
Comment (ennru) [reply compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: I kept the wording from the original, which is the same in scalac.

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 3]
```diff
@@ -1737,12 +1738,27 @@ object messages {
...
+          |Instead, you should rely on the last expression's value being
+          |returned from a method.
+          |"""
   }
 
+  case class ReturnOutsideMethodDefinition(checkedContext: Context)(implicit ctx: Context)
+    extends Message(ReturnOutsideMethodDefinitionID) {
+    val kind = "Syntax"
+    val msg = hl"${"return"} outside method definition"
+    private val contextInfo =
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Should probably be a `def`

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Ok

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Though we can improve

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 4]
```diff
@@ -980,7 +980,7 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     def enclMethInfo(cx: Context): (Tree, Type) = {
       val owner = cx.owner
       if (cx == NoContext || owner.isType) {
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala]: @nicolasstucki I'm curious. Does it ever happen to have `NoContext`?

**Comment by allanrenucci** [comment]
Can your rebase on top of master. Also you can now define the message as follow:
```scala
  case class ReturnOutsideMethodDefinition(owner: Symbol)(implicit ctx: Context)
    extends Message(ReturnOutsideMethodDefinitionID) {
    val kind = "Syntax"
    val msg = hl"${"return"} outside method definition"
    val explanation =
      hl"""You used ${"return"} in ${owner.show}".
          |${"return"} is a keyword and may only be used within method declarations.
          |"""
  }
```

**Comment by ennru** [comment]
So I understand, the `NoContext` case does not exist in practice?

**Comment by allanrenucci** [comment]
>So I understand, the NoContext case does not exist in practice?

No, `cx` should never be `NoContext`.  It was fixed in #3188

**Comment by ennru** [comment]
Sorry, looks like git and I didn't come along that well.
I'll make a new clean PR.

**Pull request closed without merging**